### PR TITLE
Custom ESLint rule: Prefer type fest

### PR DESCRIPTION
**Pull request by blazejkustra**
ISSUE: https://github.com/Expensify/App/issues/41991

### CODE REVIEW

**Review by roryabraham — CHANGES_REQUESTED** [review]
@blazejkustra can you add a test case when something else is imported from `type-fest`?

input:

```ts
import {RequireAtLeastOne} from 'type-fest';
const COLORS = { GREEN: 'green', BLUE: 'blue' } as const;
type Bad = (typeof COLORS)[keyof COLORS];
```

output:

```ts
// Note: ValueOf is added to an existing import
import {RequireAtLeastOne, ValueOf} from 'type-fest';
const COLORS = { GREEN: 'green', BLUE: 'blue' } as const;
type Bad = ValueOf<typeof COLORS>;
```

File: eslint-plugin-expensify/CONST.js [hunk 1]
```diff
@@ -30,5 +30,7 @@ module.exports = {
         USE_PERIODS_ERROR_MESSAGES: 'Use periods at the end of error messages.',
         USE_DOUBLE_NEGATION_INSTEAD_OF_BOOLEAN: 'Use !! instead of Boolean().',
         NO_ACC_SPREAD_IN_REDUCE: 'Avoid a use of spread (`...`) operator on accumulators in reduce callback. Mutate them directly instead.',
+        PREFER_TYPE_FEST_TUPLE_TO_UNION: 'Prefer using TupleToUnion from type-fest for converting tuple types to union types.',
```
Comment (fabioh8010) [inline eslint-plugin-expensify/CONST.js:33]:
```suggestion
        PREFER_TYPE_FEST_TUPLE_TO_UNION: 'Prefer using `TupleToUnion` from `type-fest` for converting tuple types to union types.',
```

File: eslint-plugin-expensify/CONST.js [hunk 2]
```diff
@@ -30,5 +30,7 @@ module.exports = {
...
+        PREFER_TYPE_FEST_VALUE_OF: 'Prefer using ValueOf from type-fest to extract the type of the properties of an object.',
```
Comment (fabioh8010) [inline eslint-plugin-expensify/CONST.js:34]:
```suggestion
        PREFER_TYPE_FEST_VALUE_OF: 'Prefer using `ValueOf` from `type-fest` to extract the type of the properties of an object.',
```

File: eslint-plugin-expensify/prefer-type-fest.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            return (fixer) => {
+                // Create replacements and add import if necessary
+                const fixes = [fixer.replaceText(node, `ValueOf<typeof ${objectName}>`)];
+    
+                if (!typeFestImported) {
+                    fixes.push(
+                        fixer.insertTextBefore(
+                            context.getSourceCode().ast.body[0],
+                            "import type {ValueOf} from 'type-fest';\n"
+                        )
```
Comment (fabioh8010) [inline eslint-plugin-expensify/prefer-type-fest.js:27]: Maybe you could extract this into a separate function (addImport), so can be used in the other use case and more in the future if we have anothers?

**Review by fabioh8010 — COMMENTED** [review]
minor comments

**Comment by blazejkustra** [comment]
@roryabraham & @fabioh8010 Adjusted the code after your suggestions, now all import patterns are covered 👍

File: eslint-plugin-expensify/utils/imports.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
+function addNamedImport(context, fixer, importNode, importName, importPath, importAsType = false) {
+  const fixes = [];
+  
+  if (importNode) {
+      const alreadyImported = importNode.specifiers.some(
+          specifier => specifier.imported.name === importName
+      );
+
+      if(!alreadyImported) {
```
Comment (fabioh8010) [inline eslint-plugin-expensify/utils/imports.js:20]:
```suggestion
      if (!alreadyImported) {
```

**Review by fabioh8010 — APPROVED** [review]
minor comment, LGTM!

**Review by roryabraham — APPROVED** [review]

**Pull request merged**